### PR TITLE
HPCC-32982 Add alternative span scopes & OwnedSpanScope refactoring

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added ActiveSpanScope to track only current active span
- Renamed existing OwnedSpanScope to OwnedActiveSpanScope
- Added OwnedSpanScope to control only span lifetime

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [X] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [X] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [X] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [X] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jtrace.hpp [hunk 1]
```diff
@@ -159,11 +159,70 @@ interface ISpan : extends IInterface
     virtual const char* queryLocalId() const = 0;
 };
 
+//------------------------------------------------------------------------------
```
Comment (jpmcmu) [inline system/jlib/jtrace.hpp:162]: I felt ActiveSpanScope and OwnedActiveSpanScope better explained what these classes are doing but I am certainly open to changing the names.

File: system/jlib/jtrace.hpp [hunk 2]
```diff
@@ -159,11 +159,70 @@ interface ISpan : extends IInterface
...
+    ISpan * query() const { return span; }
+    void set(ISpan * _span);
+    void setown(ISpan * _span);
+
+private:
+    Owned<ISpan> span;
+    ISpan * prevSpan = nullptr;
+};
+
 class jlib_decl OwnedSpanScope
```
Comment (ghalliday) [inline system/jlib/jtrace.hpp:221]: Suggest reordering this before OwnedActiveSpanScope, so there is a logical progression of functionality within the classes.

File: system/jlib/jtrace.cpp [hunk 3]
```diff
@@ -1504,13 +1504,36 @@ ISpan * CTraceManager::createServerSpan(const char * name, const IProperties * h
...
+{
+    setThreadedActiveSpan(_ptr);
+}
+
+ActiveSpanScope::~ActiveSpanScope()
+{
+    ISpan* current = queryThreadedActiveSpan();
+    if (current != span)
+    {
+        const char* currSpanID = current != nullptr ? current->querySpanId() : "null";
```
Comment (ghalliday) [inline system/jlib/jtrace.cpp:1518]: minor: queryThreadedActiveSpan() can never be null, but CNullSpan::querySpanId() should be changed to return "null".

File: system/jlib/jtrace.hpp [hunk 4]
```diff
@@ -159,11 +159,70 @@ interface ISpan : extends IInterface
...
 };
 
+//------------------------------------------------------------------------------
+// ActiveSpanScope vs OwnedActiveSpanScope Usage:
+//------------------------------------------------------------------------------
+// The primary difference between OwnedActiveSpanScope and ActiveSpanScope is that
+// OwnedActiveSpanScope controls the lifetime of its ISpan while ActiveSpanScope
+// does not. In cases where the ISpan will be used from a single thread and within
+// a single scope OwnedActiveSpanScope should be used. For more complicated scenarios,
+// involving multiple threads, time sliced work, etc ActiveSpanScope should be used.
```
Comment (ghalliday) [inline system/jlib/jtrace.hpp:169]: Add an extra comment?  (Do you agree with it?)

involving multiple threads,  time sliced work, ...

a OwnedSpanScope should be stored as a member in a class, and an ActiveSpanScope used to associate that span with each processing thread.

File: system/jlib/jtrace.hpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jtrace.hpp:221]: I almost wonder if OwnedSpanLifetime would be a better name - since it isn't really associated with a scope.

**Review by ghalliday — COMMENTED** [review]
@jpmcmu looks good.  A few minor comments.

**Review by ghalliday — APPROVED** [review]
@jpmcmu that looks good.  This will be a really helpful distinction.  I think there is one place in esp that should be revisited now these classes exist - I'll create a jira to cover that.
Please squash and I will merge.

**Comment by jpmcmu** [comment]
@ghalliday squashed

**Pull request merged**